### PR TITLE
✨ Use `pull_request_target` + protected env for e2e

### PR DESCRIPTION
**Pull request by azeemshaikh38**
* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] PR title follows the guidelines defined in  https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: .github/workflows/integration.yml
Comment (laurentsimon) [inline .github/workflows/integration.yml:38]: why do we need this if the `pull_request` trigger is not defined for this workflow?

**Review by laurentsimon — COMMENTED** [review]
did you try using `workflow_run` as suggested by someone? Why can we not use it?

File: .github/workflows/integration.yml
Comment (azeemshaikh38) [reply .github/workflows/integration.yml]: Done.

**Comment by azeemshaikh38** [comment]
I can't get the integration test workflow to run in this PR. How do we test changes to such workflows?

**Comment by laurentsimon** [comment]
> I can't get the integration test workflow to run in this PR. How do we test changes to such workflows?

the code used by GitHub is the one in the target/base branch. You'll have to merge into main to test it out. Maybe try it on a cloned repo?

File: .github/workflows/integration.yml
Comment (asraa) [inline .github/workflows/integration.yml:33]: doesn't the approve job always succeed? confused what approval is given

Comment (azeemshaikh38) [reply .github/workflows/integration.yml:33]: I should probably name that job better. `approve` job here just prints a message (I was testing whether `integration` workflow actually runs). The real approval comes from the `environment: integration-test` line. That is a protected environment which can only be accessed by PRs that have approval from one of the required approvers.

**Comment by azeemshaikh38** [comment]
> > I can't get the integration test workflow to run in this PR. How do we test changes to such workflows?
> 
> the code used by GitHub is the one in the target/base branch, IFAIK. You'll have to merge into main to test it out. Maybe try it on a cloned repo?

So I setup my fork to hav this change - https://github.com/azeemshaikh38/scorecard/blob/main/.github/workflows/integration.yml. Could you send my fork a PR from your fork to test this?

**Comment by laurentsimon** [comment]
> > > I can't get the integration test workflow to run in this PR. How do we test changes to such workflows?
> > 
> > 
> > the code used by GitHub is the one in the target/base branch, IFAIK. You'll have to merge into main to test it out. Maybe try it on a cloned repo?
> 
> So I setup my fork to have this change - https://github.com/azeemshaikh38/scorecard/blob/main/.github/workflows/integration.yml. Could you send my fork a PR from your fork to test this?

you can create PRs to you fork, right? Does it make a difference if I send the PR or you do?
Want to be sure you can iterate on sending PRs for testing, not be blocked by me to resend, etc

**Comment by azeemshaikh38** [comment]
> you can create PRs to you fork, right? Does it make a difference if I send the PR or you do? I want to be sure you can iterate on sending more PRs for testing, and not be blocked by me to resend, etc

You are right I was able to test it out - 
image

I think this PR is good to submit.

File: .github/workflows/integration.yml
Comment (laurentsimon) [inline .github/workflows/integration.yml:24]: how does this check a reviewer approved? I see only a printed message.

**Review by laurentsimon — APPROVED** [review]
LGTM overall. I don't think this is the right approach because it affects productivity so much. Every time we submit a PR, we'll have to wait for someone to LGTM to run the integration test. 

As an interim solution, it's better than having PR blocked for ever so let's merge in.

I think the right approach is to use `workflow_run`.

wdut?

**Review by laurentsimon — CHANGES_REQUESTED** [review]
see comment about printed message. I don't understand how we know it's been reviewed.

**Comment by azeemshaikh38** [comment]
> see comment about printed message. I don't understand how we know it's been reviewed.

See https://github.com/ossf/scorecard/pull/1308#discussion_r753348139

File: .github/workflows/integration.yml
Comment (azeemshaikh38) [reply .github/workflows/integration.yml:24]: It doesn't it only prints a message as you mentioned. The approve happens because of `environment` field.

Comment (laurentsimon) [reply .github/workflows/integration.yml:24]: Thanks for clarifying.

**Review by laurentsimon — APPROVED** [review]
Thanks Azeem. Let's merge this one in and merge all the blocked PRs. Do you agree that we should move towards `workflow_run` to avoid affecting productivity? (In a follow-up PR)

**Comment by azeemshaikh38** [comment]
> Thanks Azeem. Let's merge this one in and merge all the blocked PRs. Do you agree that we should move towards `workflow_run` to avoid affecting productivity? (In a follow-up PR)

No `workflow_run` will not work for us. At least, AFAICT. We can build the binaries in `pull_request` and pass it to `workflow_run` so that it can use the secret token, but then the binary itself will be a security threat.

**Comment by azeemshaikh38** [comment]
This is still stuck :( I'm going to try and force submit it. Hopefully it should fix the issue for the other PRs.

**Comment by laurentsimon** [comment]
> > Thanks Azeem. Let's merge this one in and merge all the blocked PRs. Do you agree that we should move towards `workflow_run` to avoid affecting productivity? (In a follow-up PR)
> 
> No `workflow_run` will not work for us. At least, AFAICT. We can build the binaries in `pull_request` and pass it to `workflow_run` so that it can use the secret token, but then the binary itself will be a security threat.

I see, that makes sense. Let's see if it's usable or not. Otherwise let's move to a model where we run the e2e tests without admin token. My understanding is that the token requires admin for certain check like branch protection: is my understanding correct? If that's the case, maybe we could at least separate the branch protection from the rest, to avoid blocking PR authors.

@naveensrinivasan has allowed/approved many PRs to unblock us, but it was not actually a code review: there was no LGTM. What I'm concerned with is that we're going to approve in a rush to allow the tests to run, without doing a proper review. @naveensrinivasan how is the user experience to approve these test ti run?
I'll try to go thru it on my end as well.

**Comment by naveensrinivasan** [comment]
TBH this is painful!

**Pull request merged**
